onnxruntime/core/providers/rocm/rocm_execution_provider.cc

```diff
@@ -404,10 +404,13 @@ Status ROCMExecutionProvider::EnqueueDeferredRelease() {
     // we should replace the following line with
     //  hipLaunchHostFunc(stream, ReleaseCpuBufferCallback, cpu_buffers_info);
     if (cpu_buffers_info->allocator->Info().alloc_type == OrtArenaAllocator) {
-      // Release memory asynchronously to avoid blocking the compute stream.
-      HIP_RETURN_IF_ERROR(hipStreamAddCallback(stream, ReleaseCpuBufferCallback, cpu_buffers_info.release(), 0));
+      if (actually_defer) {
+        // Release memory asynchronously to avoid blocking the compute stream.
+        HIP_RETURN_IF_ERROR(hipStreamAddCallback(stream, ReleaseCpuBufferCallback, cpu_buffers_info.release(), 0));
+      } else {
+        ReleaseCpuBufferCallback(nullptr, hipSuccess, cpu_buffers_info.release());
```

Comment:
>cpu_buffers_info [](http://example.com/codeflow?start=54&length=16)

A general question of "cpu_buffer_info", it gets released in one of multiple places on as function argument. My understanding is that with "happy-path", the pointer will be passed to another owner and get released properly later, but what if the function failed somewhere before the "owner" properly take over the ownership? For example, are we totally safe when hipStreamAddCallback returns error?

Reply:
That's an interesting observation. In the current implementation, I don't see how we're safe if/when `hipStreamAddCallback` returns an error and the callback isn't enqueued. In that case, I suspect that we'd lose track of the memory pointed to by `cpu_buffers_info`

Reply:
If it is an unrecoverable lost, maybe `HIP_CALL_THROW` it?

Reply:
@cloudhan: Done!

Reply:
Actually, on second thought, I'm not sure that `HIP_CALL_THROW` is the right behavior there. What happens if the `hipStreamEnqueueCallback` fails? We have a memory leak, which may not affect the user at all, in practice. So aborting isn't really the right behavior here, continuing execution sounds like a better idea.

Reply:
Then a FATAL log to notice the user will be sufficient. I dont think there will be profiling in long running session, then the leaking might just be harmless, since the OS will to the garbarge collect ;)